src/cylinder.jl

```diff
@@ -25,12 +25,21 @@ function Base.in(x::SVector{N}, s::Cylinder{N}) where {N}
     return sum(abs2,d - p*s.a) ≤ s.r^2
 end
 
-function normal(x::SVector{N}, s::Cylinder{N}) where {N}
+function surfpt_nearby(x::SVector{N}, s::Cylinder{N}) where {N}
     d = x - s.c
-    p = dot(d, s.a)
-    p > s.h2 && return s.a
-    p < -s.h2 && return -s.a
-    return normalize(d - p*s.a)
+    p = d ⋅ s.a
+    q = d - p*s.a
+    lp = abs(p)
+    lq = norm(q)
+    if abs(lp-s.h2) < abs(lq-s.r)
+        nout = sign(p)*s.a
+        l∆x = s.h2 - lp
+    else
+        l∆x = s.r - lq
+        nout = q / lq
```

Comment:
In fact, if I change this line to `nout = lq == 0 ? s.a : q /lq`, then `l∆x` in the line above must be changed as well, so I took a [different route](https://github.com/stevengj/GeometryPrimitives.jl/pull/17/commits/3e94ec4721c2d0af44c288bc2e4790ce61e46d41#diff-053741ac6fda59afcef4c09d24cefd77R34) in 3e94ec4.